File: lib/features/map/domain/repositories/map_repository.dart
import 'package:fpdart/fpdart.dart';
import 'package:station_reach/features/map/domain/enums/transit_mode.dart';
import 'package:station_reach/features/map/domain/models/departure.dart';
import 'package:station_reach/features/map/domain/models/station.dart';
import 'package:webfabrik_theme/webfabrik_theme.dart';

abstract class MapRepository {
  /// Searches for stations by name.
  ///
  /// Parameters:
  /// - query: Query [String] to search for
  ///
  /// Returns:
  /// - [List] of [Station]s found
  ///
  /// Failures:
  /// {@macro converted_dio_exceptions}
  Future<Either<Failure, List<Station>>> searchStations({
    required String query,
  });

  /// Gets the departures for a station by mode.
  ///
  /// Parameters:
  /// - station: [Station] to get the departures for
  /// - modes: [List] of [TransitMode]s to get the departures for
  /// - amount: [int] number of departures to get
  ///
  /// Returns:
  /// - [List] of [Departure]s found
  ///
  /// Failures:
  /// - [NoDeparturesFoundFailure]
  /// {@macro converted_dio_exceptions}
  Future<Either<Failure, List<Departure>>> getStationDeparturesByMode({
    required Station station,
    required List<TransitMode> modes,
    required int amount,
  });
}

File: src/features/map/domain/repositories/map-repository.ts
import type { Result } from 'neverthrow';

import type { Failure } from '@/core/failures';
// eslint-disable-next-line @typescript-eslint/no-unused-vars -- {@link} target
import type { NetworkingFailure } from '@/core/failures/networking-failures';

import type { Departure } from '../models/departure';
import type { Station } from '../models/station';
import type { TransitMode } from '../models/transit-mode';

/** One departures request, narrowed to a single bucket of transit modes. */
export interface DeparturesQuery {
  readonly station: Station;
  readonly modes: readonly TransitMode[];
  /** Upper bound on how many departures the answer may contain. */
  readonly amount: number;
}

/** Reads stations and their departures. */
export interface MapRepository {
  /**
   * Searches stations by name.
   *
   * @param query - Free text the user typed.
   * @param signal - Aborts the search.
   * @returns The matching stations, or a {@link NetworkingFailure}
   */
  searchStations(
    query: string,
    signal?: AbortSignal
  ): Promise<Result<Station[], Failure>>;

  /**
   * Reads the departures leaving a station, each carrying every stop it
   * reaches and how long it takes to get there.
   *
   * @returns The departures found, empty when the station has nothing
   * scheduled for these modes, or a {@link NetworkingFailure}
   */
  getStationDeparturesByMode(
    query: DeparturesQuery
  ): Promise<Result<Departure[], Failure>>;
}
